File: KursovaLibrary/KursovaLibrary/BlockClusterTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace KursovaLibrary
{
    public class BlockClusterTree
    {
        public static bool IsAdmissible(ClusterTree s, ClusterTree t)
        {
            int p = (int)Math.Pow(2, s.level);
            double[] podil = new double[p + 1];
            for (int i = 0; i < podil.Length; i++)
            {
                podil[i] = i / (double)p;
            }
            double a = podil[s.numberOfLeaf];
            double b = podil[s.numberOfLeaf + 1];
            double c = podil[t.numberOfLeaf];
            double d = podil[t.numberOfLeaf + 1];
            double diam = b - a;
            double dist = c - b;
            if (a > c)
            {
                dist = a - d;
            }
            else
            {
                dist = c - b;
            }
            if (dist < 0)
            {
                dist = 0;
            }
            if (diam <= dist)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        private static Rkmatrix buildRkmatrix(ClusterTree t, ClusterTree s, int n)
        {
            Rkmatrix rkmatrix = new Rkmatrix();
            rkmatrix.rows = t.leaf.Length;
            rkmatrix.cols = s.leaf.Length;
            rkmatrix.k = rkmatrix.rows;
            rkmatrix.a = new double[rkmatrix.rows * rkmatrix.cols];
            rkmatrix.b = new double[rkmatrix.rows * rkmatrix.cols];
            //filling Rkmatrix
            double x0 = ((double)t.leaf[0] + 0.5 * (double)t.leaf.Length) * (1.0 / (double)n);
            int m = 0;
            for (int i = 0; i < t.leaf.Length; i++)
            {
                for (int v = 1; v < rkmatrix.k; v++)
                {
                    rkmatrix.a[m] = FunctionsBEM.amatr(t.leaf[i], n, x0, v);
                    m++;
                }
            }
            m = 0;
            for (int i = 0; i < s.leaf.Length; i++)
            {
                for (int v = 1; v < rkmatrix.k; v++)
                {
                    if (v == 0)
                    {
                        rkmatrix.b[m] = FunctionsBEM.bmatr2(s.leaf[i], n, x0, v);
                      
                        m++;
                    }
                    else
                    {
                        rkmatrix.b[m] = FunctionsBEM.bmatr1(s.leaf[i], n, x0, v);
                       
                        m++;
                    }
                }
            }

            for (int i = 0; i < rkmatrix.a.Length; i++)
            {
                rkmatrix.a[i] = -0.25;
                rkmatrix.b[i] = 1;
            }
            //end of filling Rkmatrix
            return rkmatrix;
        }

        public static Supermatrix BuildBlockClusterTree(ClusterTree t,ClusterTree s,Supermatrix spr,int n)
        {
            if (IsAdmissible(t, s))
            {
                spr.supermatrix = null;
                spr.rkmatrix = buildRkmatrix(t, s, n);
                return spr;
            }
            else
            {
                if (t.leaf.Length != 1)
                {
                    spr.rows = n;
                    spr.cols = n;
                    spr.blockrows = 2;
                    spr.blockcols = 2;
                    spr.supermatrix = new Supermatrix[spr.blockrows, spr.blockcols];
                    for (int i = 0; i < spr.supermatrix.GetLength(0); i++)
                    {
                        for (int j = 0; j < spr.supermatrix.GetLength(1); j++)
                        {
                            spr.supermatrix[i, j] = new Supermatrix();
                        }
                    }
                    spr.supermatrix[0, 0] = BuildBlockClusterTree(t.leftTree, s.leftTree, spr.supermatrix[0, 0],n);
                    spr.supermatrix[0, 1] = BuildBlockClusterTree(t.rightTree, s.leftTree, spr.supermatrix[0, 1],n);
                    spr.supermatrix[1, 0] = BuildBlockClusterTree(t.leftTree, s.rightTree, spr.supermatrix[1, 0],n);
                    spr.supermatrix[1, 1] = BuildBlockClusterTree(t.rightTree, s.rightTree, spr.supermatrix[1, 1],n);
                }
                else
                {
                    spr.rows = n;
                    spr.cols = n;
                    spr.supermatrix = null;
                    spr.fullmatrix = new Fullmatrix();
                    spr.fullmatrix.cols = 0;
                    spr.fullmatrix.rows = 0;
                    spr.fullmatrix.e = new double[1];
                    //filling Fullmatrix
                    spr.fullmatrix.e[0] = FunctionsBEM.Egtulda(t.leaf[0],s.leaf[0], n);
                    //spr.fullmatrix.e[0] = 1;
                }
                return spr;
            }
            
        }

        static public double[] MultHMatrixByVector(Supermatrix spr, double[] vct)
        {
            if (spr.supermatrix != null)
            {
                int n=vct.Length;
                double[] vct1=new double[(int)(n/2)];
                double[] vct2=new double[(int)(n/2)];
                for (int i=0;i<n/2;i++){
                    vct1[i]=vct[i];
                    vct2[i]=vct[i+(int)(n/2)];
                }
                double[] a1=MultHMatrixByVector(spr.supermatrix[0, 0], vct1);
                double[] a2=MultHMatrixByVector(spr.supermatrix[0, 1], vct2);
                double[] b1=MultHMatrixByVector(spr.supermatrix[1, 0], vct1);
                double[] b2=MultHMatrixByVector(spr.supermatrix[1, 1], vct2);
                double[] res = new double[n];
                for (int i = 0; i < (int)n / 2; i++)
                {
                    res[i] = a1[i] + a2[i];
                    res[i + (int)n / 2] = b1[i] + b2[i];
                }
                return res;
            }
            else
            {
                if (spr.rkmatrix != null)
                {
                    double[,] tempa=new double[spr.rkmatrix.rows,spr.rkmatrix.cols];
                    double[,] tempb=new double[spr.rkmatrix.rows,spr.rkmatrix.cols];
                    int k = 0;
                    for (int i = 0; i < spr.rkmatrix.rows; i++)
                    {
                        for (int j = 0; j < spr.rkmatrix.cols; j++)
                        {
                            tempa[i, j] = spr.rkmatrix.a[k];
                            tempb[i, j] = spr.rkmatrix.b[k];
                            k++;
                        }
                    }
                    double[] first = GradientMethod.MultiplyMatrixByVector(tempb, vct);
                    double[] second = GradientMethod.MultiplyMatrixByVector(tempa, first);
                    return second;
                }
                else if (spr.fullmatrix != null)
                {
                    double[] res = new double[1];
                    res[0]=GradientMethod.MultiplyVectorByVector(spr.fullmatrix.e, vct);
                    return res;
                }
            }
            double[] r = new double[1];
            r[0] = 0;
            return r;
        }
    }
}

File: CouseWork/src/BlockClusterTree.kt
import kotlin.math.sign


class BlockClusterTree {
    companion object {
        //diam<=dist
        fun isAdmissible(s: ClusterTree?, t: ClusterTree?): Boolean {
            if(s!= null && t!= null) {
                val diam = 1 / Math.pow(2.0, s.level.toDouble())
                val dist = (Math.abs(s.numberOfLeaf - t.numberOfLeaf) - 1) * diam
                print("{")
                t.leaf.forEach { print(it.toString()+" ") }
                print("}x{")
                s.leaf.forEach { print(it.toString()+" ") }
                println("} - "+(diam <= dist).toString())
                return diam <= dist
            }else
                return false

        }

        fun buildRkmatrix(t: ClusterTree, s: ClusterTree, n: Int,nMin:Int): Rkmatrix {
            val k=64//2*nMin
            val rkmatrix = Rkmatrix(k, t.leaf.size, s.leaf.size)
//            //filling Rkmatrix
            val a = t.leaf[0]*(1.0 / n.toDouble())
            val b = (t.leaf[0] + t.leaf.size)*(1.0 / n.toDouble())
            val x0 = (a+b)*0.5//(t.leaf[0].toDouble() + 0.5 * t.leaf.size.toDouble()) * (1.0 / n.toDouble())
            for (i in 0 until t.leaf.size) {
                for (v in 0 until rkmatrix.k) {
                    rkmatrix.a[i][v] = amatr(t.leaf[i], n, x0, v)
                }
            }
            for (j in 0 until s.leaf.size) {
                for (v in 0 until rkmatrix.k) {
                    if (v == 0) {
                        rkmatrix.b[j][v] = bmatr2(s.leaf[j], n, x0, v)

                    } else {
                        rkmatrix.b[j][v] = bmatr1(s.leaf[j], n, x0, v)
                    }
                }
            }
            //TODO remove
//            for(i in 0 until rkmatrix.a.size){
//                for (j in 0 until rkmatrix.a[i].size){
//                    rkmatrix.a[i][j] = 2.0
//                }
//            }
//            for(i in 0 until rkmatrix.b.size){
//                for (j in 0 until rkmatrix.b[i].size){
//                    rkmatrix.b[i][j] = 5.0
//                }
//            }
            //end of filling Rkmatrix
            return rkmatrix
        }

        fun buildBlockClusterTree(t: ClusterTree?, s: ClusterTree?, spr: Supermatrix, n: Int,nMin: Int): Supermatrix {
            //admissible -rkmatrix
            if (isAdmissible(t, s)) {
                spr.supermatrix = null
                if(t!= null && s!= null)
                    spr.rkmatrix = buildRkmatrix(t, s, n,nMin)
                return spr
            } else {
                if (t?.leaf?.size!!>nMin && s?.leaf?.size!! > nMin)
                {
                    spr.rows = n
                    spr.cols = n
                    spr.blockrows = 2
                    spr.blockcols = 2
                    spr.supermatrix = Array(2){ Array(2){Supermatrix()} }
                    spr.supermatrix!![0][0] = buildBlockClusterTree(t?.leftTree, s?.leftTree, spr.supermatrix!![0][ 0],n,nMin)
                    spr.supermatrix!![0][1] = buildBlockClusterTree(t?.rightTree, s?.leftTree, spr.supermatrix!![0][1],n,nMin)
                    spr.supermatrix!![1][0] = buildBlockClusterTree(t?.leftTree, s?.rightTree, spr.supermatrix!![1][0],n,nMin)
                    spr.supermatrix!![1][1] = buildBlockClusterTree(t?.rightTree, s?.rightTree, spr.supermatrix!![1][1],n,nMin)
                }
                else
                {
                    spr.rows = n;
                    spr.cols = n;
                    spr.supermatrix = null;
                    spr.fullmatrix = Fullmatrix()
                    spr.fullmatrix!!.cols = s!!.leaf.size
                    spr.fullmatrix!!.rows = t!!.leaf.size
                    spr.fullmatrix!!.e = Array(spr.fullmatrix!!.rows){ DoubleArray(spr.fullmatrix!!.cols)}
                    //filling Fullmatrix
                    for(i in 0 until spr.fullmatrix!!.rows){
                        for (j in 0 until spr.fullmatrix!!.cols)
                            //I am not sure what is supposed to be here
                            spr.fullmatrix!!.e[i][j] = Math.abs(Egtulda(t.leaf[i],s.leaf[j], n))
                    }
                }
                return spr
            }
        }

        fun MultHMatrixByVector(spr: Supermatrix, vct: DoubleArray): DoubleArray {
            if (spr.supermatrix != null) {
                val n = vct.size
                val vct1: DoubleArray = DoubleArray(Integer.valueOf(n / 2))
                val vct2: DoubleArray = DoubleArray(Integer.valueOf(n / 2))
                for (i in 0..(n / 2 - 1)) {
                    vct1[i] = vct[i]
                    vct2[i] = vct[i + Integer.valueOf(n / 2)]
                }
                val a1 = MultHMatrixByVector(spr.supermatrix!![0][0], vct1)
                val a2 = MultHMatrixByVector(spr.supermatrix!![0][1], vct2)
                val b1 = MultHMatrixByVector(spr.supermatrix!![1][0], vct1)
                val b2 = MultHMatrixByVector(spr.supermatrix!![1][1], vct2)
                val res: DoubleArray = DoubleArray(Integer.valueOf(n))
                for (i in 0..(n / 2 - 1)) {
                    res[i] = a1[i] + a2[i]
                    res[i + Integer.valueOf(n / 2)] = b1[i] + b2[i]
                }
                return res;
            } else {
                if (spr.rkmatrix != null) {
                    val first = multiplyMatrixByVector(transposeMatrix(spr.rkmatrix!!.b), vct)
                    val second = multiplyMatrixByVector(spr.rkmatrix!!.a, first)
                    return second
                } else if (spr.fullmatrix != null) {
                    val res= multiplyMatrixByVector(spr.fullmatrix!!.e, vct)
                    return res;
                }
            }
            return DoubleArray(1)
        }
        public fun getNormalMatrix(spr:Supermatrix):Array<DoubleArray>{
            if(spr.supermatrix != null){
                val a1 = getNormalMatrix(spr.supermatrix!![0][0])
                val a2 = getNormalMatrix(spr.supermatrix!![0][1])
                val b1 = getNormalMatrix(spr.supermatrix!![1][0])
                val b2 = getNormalMatrix(spr.supermatrix!![1][1])
                val array = Array(a1.size+b1.size){DoubleArray(a1[0].size+a2[0].size)}
                for (i in 0 until a1.size)
                    for (j in 0 until a1[0].size)
                        array[i][j] = a1[i][j]
                for (i in 0 until a1.size)
                    for (j in a1.size until (a1[0].size+a2[0].size))
                        array[i][j] = a2[i][j-a1.size]
                for (i in a1.size until (a1.size+b1.size))
                    for (j in a1.size until (a1[0].size+a2[0].size))
                        array[i][j] = b2[i-a1.size][j-a1.size]
                for (i in a1.size until (b1.size+a1.size))
                    for (j in 0 until a1[0].size)
                        array[i][j] = b1[i-a1.size][j]
                return array
            }
            else {
                if (spr.rkmatrix != null) {
                    val temp = multiplyMatricesV3(spr.rkmatrix!!.a,transposeMatrix(spr.rkmatrix!!.b))
                    return multiplyMatricesV3(spr.rkmatrix!!.a,transposeMatrix(spr.rkmatrix!!.b))
                } else if (spr.fullmatrix != null) {
                    return spr.fullmatrix!!.e
                }
            }
            return Array(1){ DoubleArray(1) }
        }
    }

}